storage/remote/azuread/token_provider.go

```diff
@@ -0,0 +1,146 @@
+// limitations under the License.
+
+package azuread
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+const (
+	// Clouds
+	AzureChina      = "AzureChina"
+	AzureGovernment = "AzureGovernment"
+	AzurePublic     = "AzurePublic"
+
+	// Audiences
+	INGESTION_CHINA_AUDIENCE      = "https://monitor.azure.cn//.default"
+	INGESTION_GOVERNMENT_AUDIENCE = "https://monitor.azure.us//.default"
+	INGESTION_PUBLIC_AUDIENCE     = "https://monitor.azure.com//.default"
+)
+
+// tokenProvider is used to store and retrieve Azure AD accessToken
+type tokenProvider struct {
+	// token is member used to store the current valid accessToken
+	token string
+	ctx   context.Context
+	// refreshDuration is used to store the refresh time duration of the current valid accessToken
+	refreshDuration time.Duration
+	// credentialClient is the Azure AD credential client that is being used to retrive accessToken
+	credentialClient azcore.TokenCredential
+	options          *policy.TokenRequestOptions
+}
+
+// TokenProvider is the interface for Credential client
+type TokenProvider interface {
+	getAccessToken() string
+}
+
+// newTokenProvider helps to fetch accessToken for different types of credential. This also takes care of
+// refreshing the accessToken before expiry. This accessToken is attached to the Authorization header while making requests.
+func newTokenProvider(ctx context.Context, cfg *AzureADConfig, cred azcore.TokenCredential) (TokenProvider, error) {
+	audience, err := getAudience(cfg.Cloud)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenProvider := &tokenProvider{
+		ctx:              ctx,
+		credentialClient: cred,
+		options:          &policy.TokenRequestOptions{Scopes: []string{audience}},
+	}
+
+	err = tokenProvider.setAccessToken()
+	if err != nil {
+		return nil, errors.New("Failed to get access token: " + err.Error())
+	}
+
+	go tokenProvider.periodicallyRefreshClientToken()
+	return tokenProvider, nil
+}
+
+// Returns the current valid accessToken
+func (tokenProvider *tokenProvider) getAccessToken() string {
+	return tokenProvider.token
+}
+
+// Retrieves a new accessToken and stores the newly retrieved token in the tokenProvider
+func (tokenProvider *tokenProvider) setAccessToken() error {
+	accessToken, err := tokenProvider.credentialClient.GetToken(tokenProvider.ctx, *tokenProvider.options)
+	if err != nil {
+		return err
+	}
+
+	tokenProvider.setToken(accessToken.Token)
+	tokenProvider.updateRefreshDuration(accessToken)
+	return nil
+}
+
+// Periodically looks at the token refreshDuration and calls setAccessToken when it is past the refreshDuration
+func (tokenProvider *tokenProvider) periodicallyRefreshClientToken() error {
+	for {
+		select {
+		case <-tokenProvider.ctx.Done():
+			return nil
+		case <-time.After(tokenProvider.refreshDuration):
+			err := tokenProvider.setAccessToken()
```

Comment:
Since this done in the background do we need locking inside the setters and getters?

Reply:
We are already using `sync/atomic`, which has locking in place and handles the race condition,  to update the values of the accessToken in the `tokenProvider` struct.

```
// setToken handles storing of accessToken value in tokenProvider.
func (tokenProvider *tokenProvider) setToken(token string) {
	var V atomic.Value
	V.Store(token)
	tokenProvider.token = V.Load().(string)
}
```